### PR TITLE
i1312: Support date and timestamp types in smv histogram helper functions

### LINKED ISSUE #1312: Allow smv histogram functions to support date and timestamp types
Currently smv histogram functions only allow string, numeric and boolean types. We should also support date and timestamp types.

```
  private def createHist(histCol: edd.Hist) = {
    val s = df(histCol.colName)
    df.schema(histCol.colName).dataType match {
      case StringType =>
        if (histCol.sortByFreq) edd.StringByFreqHistogram(s)
        else edd.StringByKeyHistogram(s)
      case _: NumericType => edd.BinNumericHistogram(s, histCol.binSize)
      case BooleanType    => edd.BooleanHistogram(s)
      case t              => throw new SmvUnsupportedType(s"data type: ${t} is not supported")
    }
  }
```

### PR DESCRIPTION
**Pull request by yw-yang**
Fix #1312 

screen shot 2018-07-19 at 1 50 14 pm

### CODE REVIEW

**Comment by laneb** [comment]
@yw-yang Is your git configured with your github email address? Your commits are stamped with your name but no email or avatar.

**Comment by yw-yang** [comment]
@laneb I just made some changes, are you able to see email or avatar now?

File: src/main/scala/org/tresamigos/smv/edd/EddTask.scala [hunk 1]
```diff
@@ -206,12 +206,12 @@ private[smv] case class BooleanHistogram(col: Column) extends EddHistTask {
 private[smv] case class StringByKeyHistogram(col: Column) extends EddHistTask {
   override val taskName = "key"
   override val taskDesc = "String sort by Key"
-  override val statOp   = histStr(col)
+  override val statOp   = histStr(col.cast("string"))
```
Comment (laneb) [inline src/main/scala/org/tresamigos/smv/edd/EddTask.scala]: Needs documentation.

**Comment by laneb** [comment]
@yw-yang No, but if you changed your configuration we should see the change when you push your next commit.

File: src/main/scala/org/tresamigos/smv/edd/EddTask.scala (hunk 1 above)
Comment (yw-yang) [reply src/main/scala/org/tresamigos/smv/edd/EddTask.scala]: Comments have been added.

Comment (yw-yang) [reply src/main/scala/org/tresamigos/smv/edd/EddTask.scala]: I agree doing a histogram of timestamp is rare and not very useful but still think smv should support it for the rare case and let user decide whether to use it or now.

**Comment by yw-yang** [comment]
I agree doing a histogram of timestamp is rare and not very useful but still think smv should support it for the rare case and let user decide whether to use it or now.

**Comment by laneb** [comment]
👍 
yuwei commit avatar

**Comment by laneb** [comment]
The question was whether it makes sense to do a histogram on the raw timestamp or whether the timestamps ought to be binned into larger time intervals similar to the `BinNumericHistogram`. I suppose. Does the timestamp hist feature have no use case?

**Comment by yw-yang** [comment]
@laneb I rarely do histogram on timestamp type, the use case I encountered before is to check the `system_load_ts`, which records the timestamp each row is appended to a table. In batch process, for each batch loading, the `system_load_ts` is always the same, so there won't be a large number of different values. Another case is in some ETL systems, although `system_load_ts` is timestamp type, the value just contains date information, e.g. "2018-01-01 00:00:00". Of course, user can always cast timestamp type to date/string type and do `smvHist`, but it is an additional burden and not user-friendly. 

Another concern is doing histogram on timestamp may hang the system and run out of memory. In ticket #988,  I propose we can set a threshold in `smvHist` internally, e.g. if # values is greater than 1000, method stops calculating and shows a warning instead. What do you think?

**Comment by laneb** [comment]
That makes sense. Maybe we should make the timestamp binning usecase a little more explicit? If we are expecting that timestamps are binned in order for this to work, we should probably make the binning explicit and a bit easier - e.g. enumerate a few levels of granularity that the timestamps can be binned into. Then we can also make the representation in the UI a bit nicer - truncate the timestamp string to the significant digits.

**Comment by yw-yang** [comment]
Do you mean creating another function, e.g. `smvTimeHist(*colwithGranularity)` to deal with timestamp binning like below? If so, I will open another ticket for it.
```
d.smvTimeHist(("ts1", "Year"), ("ts2", "Month"))
```

**Comment by laneb** [comment]
Yeah, it's outside of the scope of this ticket. Just seems like it would be a more useful API if we really want to support timestamp hist.

**Pull request merged**